### ISSUE TITLE
Arithmetic exception in strmNextFile

### ISSUE THREAD

**Issue by eugenepaniot** [rsyslog/rsyslog#1066]
We had got a lot of coredump's with "Arithmetic exception" message at stream.c:482

```
Core was generated by `/sbin/rsyslogd -i /var/run/syslogd.pid'.
Program terminated with signal 8, Arithmetic exception.
\#0  0x00007fc969fc751a in strmNextFile (pThis=0x7fc93000ec20) at stream.c:482
482             pThis->iCurrFNum = (pThis->iCurrFNum + 1) % pThis->iMaxFiles;
```

```
(gdb) where
#0  0x00007fc969fc751a in strmNextFile (pThis=0x7fc93000ec20) at stream.c:482
#1  0x00007fc969fc83b0 in strmSerialize (pThis=0x7fc93000e3b0, pStrm=0x7fc93000ec20) at stream.c:1988
#2  0x00007fc9673313eb in persistStrmState (pLstn=0x7fc930004c10) at imfile.c:1966
#3  0x00007fc967331466 in lstnDel (pLstn=0x7fc930004c10) at imfile.c:872
#4  0x00007fc967331b7b in in_removeFile (dirIdx=0, pLstn=0x7fc930004c10) at imfile.c:1717
#5  0x00007fc967332750 in in_processEvent () at imfile.c:1832
#6  do_inotify () at imfile.c:1885
#7  0x00007fc967332b8a in runInput (pThrd=<value optimized out>) at imfile.c:1918
#8  0x00007fc969fda5b1 in thrdStarter (arg=0x7fc96b792700) at ../threads.c:220
#9  0x00007fc9697369d1 in start_thread () from /lib64/libpthread.so.0
#10 0x00007fc96864b8fd in clone () from /lib64/libc.so.6
```

```
(gdb) p *strmNextFile::pThis
$44 = {objData = {pObjInfo = 0x7fc96b6afc10, pszName = 0x0}, sType = STREAMTYPE_FILE_SINGLE, iCurrFNum = 1, pszFName = 0x7fc930005340 "pas-main-log", lenFName = 12, tOperationsMode = STREAMMODE_WRITE_TRUNC, tOpenMode = 384, 
  iMaxFileSize = 0, iMaxFiles = 0, iFileNumDigits = 0, bDeleteOnClose = 0 '\000', iCurrOffs = 0, pUsrWCntr = 0x0, bPrevWasNL = 0 '\000', bDisabled = 0 '\000', bSync = 0 '\000', bReopenOnTruncate = 0 '\000', sIOBufSize = 4096, 
  pszDir = 0x7fc930005f30 "/var/log/spool/rsyslog", lenDir = 22, fd = -1, fdDir = -1, inode = 0, pszCurrFName = 0x0, 
  pIOBuf = 0x7fc93000d160 " ... LONG MESSAGE ... ", 
iBufPtrMax = 0, iBufPtr = 0, iUngetC = -1, bInRecord = 0 '\000', iZipLevel = 0, pZipBuf = 0x0, 
  bAsyncWrite = 0 '\000', bStopWriter = 0 '\000', bDoTimedWait = 0 '\000', bzInitDone = 0 '\000', bVeryReliableZip = 0 '\000', iFlushInterval = 0, mut = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, 
      __spins = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}, notFull = {__data = {__lock = 0, __futex = 0, __total_seq = 0, __wakeup_seq = 0, __woken_seq = 0, __mutex = 0x0, __nwaiters = 0, 
      __broadcast_seq = 0}, __size = '\000' <repeats 47 times>, __align = 0}, notEmpty = {__data = {__lock = 0, __futex = 0, __total_seq = 0, __wakeup_seq = 0, __woken_seq = 0, __mutex = 0x0, __nwaiters = 0, __broadcast_seq = 0}, 
    __size = '\000' <repeats 47 times>, __align = 0}, isEmpty = {__data = {__lock = 0, __futex = 0, __total_seq = 0, __wakeup_seq = 0, __woken_seq = 0, __mutex = 0x0, __nwaiters = 0, __broadcast_seq = 0}, 
    __size = '\000' <repeats 47 times>, __align = 0}, iEnq = 0, iDeq = 0, cryprov = 0x0, cryprovData = 0x0, cryprovFileData = 0x0, iCnt = 0, zstrm = {next_in = 0x0, avail_in = 0, total_in = 0, next_out = 0x0, avail_out = 0, 
    total_out = 0, msg = 0x0, state = 0x0, zalloc = 0, zfree = 0, opaque = 0x0, data_type = 0, adler = 0, reserved = 0}, asyncBuf = {{pBuf = 0x0, lenBuf = 0}, {pBuf = 0x0, lenBuf = 0}}, writerThreadID = 0, iSizeLimit = 0, 
  pszSizeLimitCmd = 0x0, bIsTTY = 0 '\000', prevLineSegment = 0x0, prevMsgSegment = 0x0}
(gdb) 

```

It seems some race condition, because fd is -1 and ASSERT(pThis->fd != -1) doesn't works.

**Comment by rgerhards** [comment]
Do you can give instructions on how this problem could be reproduced?

**Comment by eugenepaniot** [comment]
Not sure. Our rsyslog configurations are:

**/etc/rsyslog.conf**
```
global(
    action.reportSuspension="on"
    net.enableDNS="off"
    net.ipprotocol="ipv4-only"
    senders.keepTrack="on"
    senders.reportGoneAway="on"
    senders.reportNew="on"
)

timezone(id="UTC" offset="+00:00")

$MaxOpenFiles 65535

module(load="impstats" interval="60" bracketing="on" severity="7" facility="6" log.syslog="on" ruleset="impstats")

$ModLoad immark

## provides UDP syslog reception
module(load="imudp")
input(type="imudp" address="0.0.0.0" port="514" rcvbufSize="1m" RateLimit.Interval="0" name="udp-server")

## provides TCP syslog reception
module(load="imtcp" MaxSessions="500" FlowControl="off" MaxListeners="32" )
input(type="imtcp" address="0.0.0.0" port="514" name="tcp-server")

## provides kernel logging support.
$ModLoad imklog
$KLogPermitNonKernelFacility on

## provides read from file feature 
module(load="imfile")

## provides support for local system logging
module(load="imuxsock" SysSock.RateLimit.Interval="0" SysSock.UsePIDFromSystem="on")
input(type="imuxsock" Socket="/dev/log" RateLimit.Interval="0" UsePIDFromSystem="on" )

$WorkDirectory /var/log/spool/rsyslog

$MainMsgQueueFileName rsyslog-main-queue
$MainMsgQueueType Direct
$MainMsgQueueSaveOnShutdown on
$MainMsgQueueTimeoutShutdown 120000 # number is timeout in ms - 120sec
$MainMsgQueueTimeoutEnqueue 0 #We do not like throtteling, the message will then immediately be discarded.

$ActionQueueType Direct
$ActionQueueFileName rsyslog-disk-queue
$ActionResumeRetryCount 3
$ActionQueueMaxFileSize 100m
$ActionQueueSaveOnShutdown on

# log every message
$RepeatedMsgReduction off

# Multiline support:
$EscapeControlCharactersOnReceive off

# Increased maximum message size:
$MaxMessageSize 640k

$LogRSyslogStatusMessages on

$ActionFileDefaultTemplate RSYSLOG_RCFormat
$ActionForwardDefaultTemplate RSYSLOG_RCFormat

$template RSYSLOG_RCFormat, "%TIMESTAMP:::date-rfc3339% %HOSTNAME% %syslogseverity-text:::uppercase% %syslogtag%%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%\n"

# include additional config files
$IncludeConfig /etc/rsyslog.d/*.conf

###------------------- Permissions -------------------------------
##$umask 0000

$FileCreateMode 0600
$FileOwner root
$FileGroup root

$DirCreateMode 0755
$DirOwner root
$DirGroup root


##----------------- Logging Configuration ------------------------
# Log kernel messages
kern.*              action(name="kern" type="omfile" File="/var/log/kern.log")

*.info;mail,authpriv,cron.none      action(name="messages" type="omfile" File="/var/log/messages")

# The authpriv file has restricted access.
authpriv.*          action(name="secure" type="omfile" File="/var/log/secure")

# Log all the mail messages in one place.
mail.*              action(name="maillog" type="omfile" File="/var/log/maillog")

# Log cron stuff
cron.*              action(name="cron" type="omfile" File="/var/log/cron")

# Everybody gets emergency messages
*.emerg             action(name="emerg" type="omusrmsg" Users="*")

# Save news errors of level crit and higher in a special file.
uucp,news.crit      action(name="spooler" type="omfile" File="/var/log/spooler")

# Save boot messages also to boot.log
local7.*            action(name="boot" type="omfile" File="/var/log/boot.log")
```


**/etc/rsyslog.d/90-sumo-raw-pas.conf**
```
#pas.statistics-api.log
$InputFileBindRuleset sumologic-pas-imfile
$InputFileName /var/log/HIDDEN/pas/pas.statistics-api.log
$InputFileTag pas-statistics-api:
$InputFileStateFile pas-statistics-api
$InputFileReadMode 2
$InputFileSeverity notice
$InputFileFacility local0
$InputFilePersistStateInterval 100000
$InputRunFileMonitor

#pas.main.log
$InputFileBindRuleset sumologic-pas-imfile
$InputFileName /var/log/HIDDEN/pas/pas.main.log
$InputFileTag pas.main.log:
$InputFileStateFile pas-main-log
$InputFileReadMode 2
$InputFileSeverity notice
$InputFileFacility local1
$InputFilePersistStateInterval 100000
$InputRunFileMonitor 

#pas.statistics-sms.log
$InputFileBindRuleset sumologic-pas-imfile
$InputFileName /var/log/HIDDEN/pas/pas.statistics-sms.log
$InputFileTag pas-statistics-sms:
$InputFileStateFile pas-statistics-sms
$InputFileReadMode 2
$InputFileSeverity notice
$InputFileFacility local2
$InputFilePersistStateInterval 100000
$InputRunFileMonitor

ruleset(name="sumologic-pas-imfile"
        queue.filename="queue-sumologic-pas-imfile"
        queue.maxdiskspace="500m"
        queue.maxfilesize="10m"
        queue.type="FixedArray"
        queue.workerthreads="1"
        queue.syncqueuefiles="off"
        queue.checkpointinterval="100000"
        queue.size="10000"
        queue.discardmark="9800"
        queue.dequeuebatchsize="1024"
        queue.timeoutenqueue="0"
)
{
    
local0.*        action(type="omfwd" protocol="udp" target="sjc01-p01-lcs01.p01.HIDDEN.com" port="5050" 
      name="omfwd-sumologic-pas-imfile-udp-sjc01-p01-lcs01.p01.HIDDEN.com-5050" 
      queue.FileName="omfwd-sumologic-pas-imfile-udp-sjc01-p01-lcs01.p01.HIDDEN.com-5050" 
      template="SumoRawFormat" 
      action.resumeRetryCount="3" )

local1.*        action(type="omfwd" protocol="udp" target="sjc01-p01-lcs01.p01.HIDDEN.com" port="5051" 
      name="omfwd-sumologic-pas-imfile-udp-sjc01-p01-lcs01.p01.HIDDEN.com-5051" 
      queue.FileName="omfwd-sumologic-pas-imfile-udp-sjc01-p01-lcs01.p01.HIDDEN.com-5051" 
      template="SumoRawFormat" 
      action.resumeRetryCount="3" )

local2.*        action(type="omfwd" protocol="udp" target="sjc01-p01-lcs01.p01.HIDDEN.com" port="5052" 
      name="omfwd-sumologic-pas-imfile-udp-sjc01-p01-lcs01.p01.HIDDEN.com-5052" 
      queue.FileName="omfwd-sumologic-pas-imfile-udp-sjc01-p01-lcs01.p01.HIDDEN.com-5052" 
      template="SumoRawFormat" 
      action.resumeRetryCount="3" )

    
    *.*     call remote
}
```


**/etc/rsyslog.d/001-rsyslog-remote.conf**
```
$Modload mmsequence
$template RSYSLOG_RCTFormat, "%TIMESTAMP:::date-rfc3339% %HOSTNAME% %$!seq1% %syslogseverity-text:::uppercase%\9 %syslogtag%%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%\n"


*.*      call remote


ruleset(name="remote"
        queue.filename="queue-remote"
        queue.maxdiskspace="500m"
        queue.maxfilesize="10m"
        queue.type="FixedArray"
        queue.syncqueuefiles="off"
        queue.checkpointinterval="100000"
        queue.size="10000"
        queue.discardmark="9800"
        queue.dequeuebatchsize="1024"
        queue.timeoutenqueue="0"
)
{
    *.*   action(type="mmsequence" 
     mode="key" 
     key="key1" 
     var="$!seq1" )
action(type="omfwd" template="RSYSLOG_RCTFormat" protocol="tcp" target="logstash01.HIDDEN.com" port="10001" 
     name="omfwd-remote-logstash01.HIDDEN.com" 
     queue.FileName="omfwd-remote-logstash01.HIDDEN.com" 
     action.resumeRetryCount="3" )
}
```

So, write some messages single line and/or multiline to pas.statistics-api.log, pas.main.log, pas.statistics-sms.log about >100 per second.

**Comment by eugenepaniot** [comment]
Rsyslog version is rsyslog-8.18.0-1.el6.x86_64

**Comment by mostolog** [comment]
@eugenepaniot Could you check if this is solved with 8.22 + patch from https://github.com/rsyslog/rsyslog/issues/1239 (it's just one line!)

**Comment by rgerhards** [comment]
I missed that issue tracker. Judging from the segfault, it is fixed by https://github.com/rsyslog/rsyslog/pull/1241, so closing this.
